### DISCUSSION TITLE
Compatibility with django-debug-toolbar-template-profiler

### DISCUSSION THREAD

**Discussion by oliverhaas** [django-components/django-components#819] [Q&A]
I'm using the above package (see https://github.com/node13h/django-debug-toolbar-template-profiler), and overall I like it quite a lot and it's quite useful.

However, it does not really seem to be compatible with django-components. I guess django-components somehow hooks directly or differently into template rendering, so whatever the profiler does, it does not see the component templates.

Has anyone experience with that and has a workaround or better explanation? I will have a closer look at this if I have the time, but I thought I check in here first.

**Comment by JuroOravec** [comment]
Haven't heard of the debug toolbar nor the template profiler before, but this would be great to have support for.

Had a quick look, just a guess, but we have to monkey patch the `Template.render()` too. But what we currently do is that we REPLACE the original implementation, instead of wrapping it, so that we can apply context isolation as necessary.

See https://github.com/JuroOravec/django-components/blob/7b406e6a2d5dedc759d9537009670443f5b8fc05/src/django_components/component.py#L983

It looks like the debug profiler properly WRAPS the original implementation, see https://github.com/node13h/django-debug-toolbar-template-profiler/blob/master/template_profiler_panel/panels/template.py#L127

So @oliverhaas when you have the time to play with it, check the order in which the patches are applied. IMO things should work if we get to apply our patch first.

---

Altho, actually, since we do the patching only at the point when we are rendering the template, then my guess is that the debug profiler and it's patch probably runs first.

(Tho this depends on where the [`TemplateProfilerPanel`](https://github.com/node13h/django-debug-toolbar-template-profiler/blob/master/template_profiler_panel/panels/template.py#L58C7-L58C28) is instantiated, and I couldn't find that inside the source code, and would have to look into the source of `django-debug` to understand where and when the `Panel` classes are instantiated).

The profiler package is using some 3rd party package called `wrapt` to wrap the original implementation. I'm not familiar with it. But when using vanilla `functools.wraps()`, then if I remember correctly you can access the original function.

So some ideas to explore:

1. ~If we can detect and access the original AND if we can modify the original function WITHOUT HAVING TO REAPPLY THE DECORATORS, then inside our monkeypatching function, we could check if `Template.render()` is wrapped, and recursively access original function, and modify that.~

2. Probably simpler and clear would be to move our patching function to be applied earlier on, e.g. at the `setup()` hook in `apps.py`. 

   And in [the old place of where we applied the patching](https://github.com/JuroOravec/django-components/blob/7b406e6a2d5dedc759d9537009670443f5b8fc05/src/django_components/component.py#L1027), we could just run a check to ensure that the `Template` class has been previously pached, or raise an error, saying that "we came across a custom 
Template class that does not extend the Django's Template class" and ask them to raise an issue in django-components.

    Because that's why I originally implemented the patching only at render time, so that if someone supplies a custom Template class that behaves like Django's Template class, but does NOT subclass Django's class, it would still work (duck-typing). So for that use case, it would be good to tell people to raise an issue with us, so we can learn of their needs and act accordingly.

    And once the Template patching is happening not at render time, but up-front in `apps.py` in `setup()`, then  I think it should be simpler to figure out the configuration to apply our patching before that of https://github.com/node13h/django-debug-toolbar-template-profiler (e.g. maybe django-components will need to be set BEFORE template-profiler inside `INSTALLED_APPS`).

**Reply by oliverhaas** [reply]
Thanks for the input. I'll hopefully take a lot in the next weeks. :smile:

**Reply by JuroOravec** [reply]
I made an MR for the changes outline above in this PR https://github.com/EmilStenstrom/django-components/pull/825. 

@oliverhaas I think that once the above is merged and released, then I think django-components should work with the template-profiler. Because our patching should then be applied first.

I looked into the template profiler:

1. The profiler calls the monkeypatch when the `TemplateProfilerPanel` is instanced ([source](https://github.com/node13h/django-debug-toolbar-template-profiler/blob/master/template_profiler_panel/panels/template.py#L66C18-L78C28))

2. `TemplateProfilerPanel` is instantiated at the same time as when `DebugToolbar` is, it instantiates all the `Panel` classes ([source](https://github.com/django-commons/django-debug-toolbar/blob/main/debug_toolbar/toolbar.py#L35)).

3. And `DebugToolbar` is instantiated from within django-debug-toolbar's middleware ([source1](https://github.com/django-commons/django-debug-toolbar/blob/main/debug_toolbar/middleware.py#L88), [source2](https://github.com/django-commons/django-debug-toolbar/blob/main/debug_toolbar/middleware.py#L111))

**Reply by JuroOravec** [reply]
This changes will be available in the next release (v0.119)

**Reply by oliverhaas** [reply]
Thank you for your work! I'm still checking out whether it works exactly as expected. At a first glance I think the timings are not correct, but I really need to take a closer look. I will have some time for this at the end of the year.
